packages/playwright-test/src/plugins/webServerPlugin.ts

```diff
@@ -13,66 +13,102 @@
  * See the License for the specific language governing permissions and
  * limitations under the License.
  */
-
 import http from 'http';
 import https from 'https';
+import path from 'path';
 import net from 'net';
+
 import { debug } from 'playwright-core/lib/utilsBundle';
 import { raceAgainstTimeout } from 'playwright-core/lib/utils/timeoutRunner';
-import type { FullConfig } from './types';
 import { launchProcess } from 'playwright-core/lib/utils/processLauncher';
-import type { Reporter } from '../types/testReporter';
+
+import type { PlaywrightTestConfig, TestPlugin } from '../types';
+import type { Reporter } from '../../types/testReporter';
+
+export type WebServerPluginOptions = {
+  command: string;
+  url: string;
+  ignoreHTTPSErrors?: boolean;
+  timeout?: number;
+  reuseExistingServer?: boolean;
+  cwd?: string;
+  env?: { [key: string]: string; };
+};
+
+interface InternalWebServerConfigOptions extends Omit<WebServerPluginOptions, 'url'> {
+  setBaseURL: boolean;
+  url?: string;
+  port?: number;
+}
 
 const DEFAULT_ENVIRONMENT_VARIABLES = {
   'BROWSER': 'none', // Disable that create-react-app will open the page in the browser
 };
 
 const debugWebServer = debug('pw:webserver');
 
-type WebServerConfig = NonNullable<FullConfig['webServer']>;
-
-export class WebServer {
+export class InternalWebServerPlugin implements TestPlugin {
```

Comment:
food for thought: imagine we are to deprecate the old webServer settings. Refactoring this code to drop it would be non-trivial. On the other hand, if you make a new shiny WebServerPlugin and build a legacy plugin on top if it, removing the legacy one would be a matter or removing an extra plugin.